From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tom Dohrmann <erbse.13@gmx.de>
Date: Wed, 10 Jul 2024 13:31:34 +0200
Subject: [PATCH] add options for library paths

For some reason meson fails to find these when building with pkgsStatic
in Nix.

Since meson 1.12, find_library also checks static libraries by linking
them into a shared object, which can't work with a static-only toolchain.
So link the archives directly if a path is given.
---
diff --git a/meson.build b/meson.build
index 40e3f4e..e660ac9 100644
--- a/meson.build
+++ b/meson.build
@@ -1161,8 +1161,12 @@ zlib = dependency('zlib', required: true)
 
 libaio = not_found
 if not get_option('linux_aio').auto() or have_block
-  libaio = cc.find_library('aio', has_headers: ['libaio.h'],
-                           required: get_option('linux_aio'))
+  if get_option('linux_aio_path') != ''
+    libaio = declare_dependency(link_args: [get_option('linux_aio_path') / 'libaio.a'])
+  else
+    libaio = cc.find_library('aio', has_headers: ['libaio.h'],
+                             required: get_option('linux_aio'))
+  endif
 endif
 
 linux_io_uring_test = '''
@@ -2045,7 +2049,11 @@ if fdt_opt == 'enabled' and get_option('wrap_mode') == 'nodownload'
   fdt_opt = 'system'
 endif
 if fdt_opt in ['enabled', 'system'] or (fdt_opt == 'auto' and have_system)
-  fdt = cc.find_library('fdt', required: fdt_opt == 'system')
+  if get_option('linux_fdt_path') != ''
+    fdt = declare_dependency(link_args: [get_option('linux_fdt_path') / 'libfdt.a'])
+  else
+    fdt = cc.find_library('fdt', required: fdt_opt == 'system')
+  endif
   if fdt.found() and cc.links('''
      #include <libfdt.h>
      #include <libfdt_env.h>
diff --git a/meson_options.txt b/meson_options.txt
index a07cb47..309afbd 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -200,6 +200,7 @@ option('linux_aio', type : 'feature', value : 'auto',
        description: 'Linux AIO support')
 option('linux_io_uring', type : 'feature', value : 'auto',
        description: 'Linux io_uring support')
+option('linux_aio_path', type: 'string', value: '', description: 'Path for libaio.a')
 option('lzfse', type : 'feature', value : 'auto',
        description: 'lzfse support for DMG images')
 option('lzo', type : 'feature', value : 'auto',
@@ -322,6 +323,7 @@ option('capstone', type: 'feature', value: 'auto',
 option('fdt', type: 'combo', value: 'auto',
        choices: ['disabled', 'enabled', 'auto', 'system', 'internal'],
        description: 'Whether and how to find the libfdt library')
+option('linux_fdt_path', type: 'string', value: '', description: 'Path for libfdt.a')
 
 option('selinux', type: 'feature', value: 'auto',
        description: 'SELinux support in qemu-nbd')
